### ISSUE TITLE
Support for TeamFoundationServer / tfs

### ISSUE THREAD

**Issue by Grueslayer** [mhinz/vim-signify#177]
I am forced to use TFS at work, so here is my patch for supporting it.
TF.EXE creates diffs with context and has no option (like U0) to reduce it, so I strip it down later.

```diff
diff -urN vim-signify-master/autoload/sy/repo.vim vim-signify/autoload/sy/repo.vim
--- vim-signify-master/autoload/sy/repo.vim	2016-02-23 15:42:27.000000000 +0100
+++ vim-signify/autoload/sy/repo.vim	2016-02-23 16:47:43.671560500 +0100
@@ -85,6 +85,12 @@
   return v:shell_error ? [0, ''] : [1, diff]
 endfunction
 
+" Function: #get_diff_tfs {{{1
+function! sy#repo#get_diff_tfs() abort
+  let diff = s:run(g:signify_vcs_cmds.tfs, b:sy_info.file, 0)
+  return v:shell_error ? [0, ''] : [1, s:strip_context(diff)]
+endfunction
+
 " Function: #get_stats {{{1
 function! sy#repo#get_stats() abort
   if !exists('b:sy') || !has_key(b:sy, 'stats')
@@ -112,6 +118,7 @@
         \ 'rcs':      [g:signify_vcs_cmds.rcs,      b:sy_info.path, 0],
         \ 'accurev':  [g:signify_vcs_cmds.accurev,  b:sy_info.file, 1],
         \ 'perforce': [g:signify_vcs_cmds.perforce, b:sy_info.path, 0],
+        \ 'tfs':      [g:signify_vcs_cmds.tfs,      b:sy_info.file, 0],
         \ }
 
   for vcs in s:vcs_list
@@ -163,6 +170,80 @@
   return join(parts, a:sub)
 endfunction
 
+" Function: s:strip_context {{{1
+function! s:strip_context(context)
+  let diff = []
+  let hunk = []
+  let state = 0
+  let lines = split(a:context,"\n",1)
+  let linenr = 0
+
+  while linenr < len(lines)
+    let line = lines[linenr]
+
+    if state == 0
+      if line =~ "^@@ "
+        let tokens = matchlist(line, '^@@ -\v(\d+),?(\d*) \+(\d+),?(\d*)')
+        let old_line = str2nr(tokens[1])
+        let new_line = str2nr(tokens[3])
+        let old_count = empty(tokens[2]) ? 1 : str2nr(tokens[2])
+        let new_count = empty(tokens[4]) ? 1 : str2nr(tokens[4])
+        let state = 1
+      else
+        call add(diff,line)
+      endif
+      let linenr = linenr + 1
+    elseif state == 1
+      if line[0] == ' '
+        let old_line = old_line + 1
+        let new_line = new_line + 1
+        let old_count = old_count - 1
+        let new_count = new_count - 1
+        let linenr = linenr + 1
+      else
+        let hunk = []
+        let old_count_part = 0
+        let new_count_part = 0
+        let state = 2
+      endif
+    elseif state == 2
+      if line[0] == '-'
+        call add(hunk,line)
+        let old_count_part = old_count_part + 1
+        let linenr = linenr + 1
+      else
+        let state = 3
+      endif
+    elseif state == 3
+      if line[0] == '+'
+        call add(hunk,line)
+        let new_count_part = new_count_part + 1
+        let linenr = linenr + 1
+      else
+        call add(diff, printf("@@ -%d,%d +%d,%d @@",old_line, old_count_part, (new_count_part == 0 && new_line > 0) ? new_line - 1 : new_line, new_count_part))
+        let diff = diff + hunk
+        let hunk = []
+        let old_count = old_count - old_count_part
+        let new_count = new_count - new_count_part
+        let old_line = old_line + old_count_part
+        let new_line = new_line + new_count_part
+        let state = 1
+      endif
+    endif
+
+    if state > 0 && new_count <= 0 && old_count <= 0
+      if len(hunk) > 0
+        call add(diff, printf("@@ -%d,%d +%d,%d @@",old_line, old_count_part, (new_count_part == 0 && new_line > 0) ? new_line - 1 : new_line, new_count_part))
+        let diff = diff + hunk
+        let hunk = []
+      endif
+      let state = 0
+    endif
+  endwhile
+
+  return join(diff,"\n")."\n"
+endfunction
+
 " Variables {{{1
 let s:difftool = get(g:, 'signify_difftool', 'diff')
 if executable(s:difftool)
@@ -176,7 +257,8 @@
         \ 'cvs':      'cvs',
         \ 'rcs':      'rcsdiff',
         \ 'accurev':  'accurev',
-        \ 'perforce': 'p4'
+        \ 'perforce': 'p4',
+        \ 'tfs':      'tf'
         \ }
 else
   echomsg 'signify: No diff tool found -> no support for svn, darcs, bzr, fossil.'
@@ -186,7 +268,8 @@
         \ 'cvs':      'cvs',
         \ 'rcs':      'rcsdiff',
         \ 'accurev':  'accurev',
-        \ 'perforce': 'p4'
+        \ 'perforce': 'p4',
+        \ 'tfs':      'tf'
         \ }
 endif
 
@@ -206,6 +289,7 @@
       \ 'rcs':      'rcsdiff -U0 %f 2>%n',
       \ 'accurev':  'accurev diff %f -- -U0',
       \ 'perforce': 'p4 info '. sy#util#shell_redirect('%n') .' && env P4DIFF=%d p4 diff -dU0 %f',
+      \ 'tfs':      'tf diff -version:W -noprompt -format:Unified %f'
       \ }
 
 if exists('g:signify_vcs_cmds')
diff -urN vim-signify-master/doc/signify.txt vim-signify/doc/signify.txt
--- vim-signify-master/doc/signify.txt	2016-02-23 15:42:27.000000000 +0100
+++ vim-signify/doc/signify.txt	2016-02-19 18:53:28.469284700 +0100
@@ -63,6 +63,7 @@
     fossil
     accurev
     perforce
+    tfs
 <
 Sign explanation:~
 
@@ -123,6 +124,7 @@
                  'perforce'
                  'rcs'
                  'svn'
+                 'tfs'
 
 NOTE: This is the most important option, so read closely.
 
@@ -186,6 +188,7 @@
       \ 'rcs':      'rcsdiff -U0 %f 2>%n',
       \ 'accurev':  'accurev diff %f -- -U0',
       \ 'perforce': 'p4 info '. sy#util#shell_redirect('%n') .' && env P4DIFF=%d p4 diff -dU0 %f'
+      \ 'tfs':      'tf diff -version:W -noprompt %f',
       \ }
 <
 ------------------------------------------------------------------------------
diff -urN vim-signify-master/README.md vim-signify/README.md
--- vim-signify-master/README.md	2016-02-23 15:42:27.000000000 +0100
+++ vim-signify/README.md	2016-02-19 18:49:41.491070300 +0100
@@ -16,7 +16,7 @@
 **Features:**
 
 - supports git, mercurial, darcs, bazaar, subversion, cvs, rcs, fossil, accurev,
-  perforce
+  perforce, tfs
 - quick jumping between blocks of changed lines ("hunks")
 - apart from signs there is also optional line highlighting
 - preserves signs from other plugins

```

**Comment by Grueslayer** [comment]
Patch corrected to fix Issue #178.

**Comment by mhinz** [comment]
There is no way I can test this, but your patch seems pretty solid, so I merged it as-is.

Thank you!
